crates/flipperzero/src/input/mod.rs

```diff
@@ -0,0 +1,125 @@
+mod key;
+mod r#type;
+
+use flipperzero_sys::{self as sys, InputEvent as SysInputEvent};
+use ufmt::derive::uDebug;
+// public type alias for an anonymous union
+pub use sys::InputEvent__bindgen_ty_1 as SysInputEventSequence;
+
+pub use key::*;
+pub use r#type::*;
+
+/// Input event occurring on user actions.
+///
+/// Corresponds to raw [`SysInputEvent`].
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub struct InputEvent {
+    /// Sequence qualifying the event.
+    pub sequence: InputEventSequence,
+    /// Physical key causing the event.
+    pub key: InputKey,
+    /// The type of the event.
+    pub r#type: InputType,
+}
+
+impl TryFrom<SysInputEvent> for InputEvent {
+    type Error = FromSysInputEventError;
+
+    fn try_from(value: SysInputEvent) -> Result<Self, Self::Error> {
+        Ok(Self {
+            sequence: value.__bindgen_anon_1.into(),
+            key: value.key.try_into()?,
+            r#type: value.type_.try_into()?,
+        })
+    }
+}
+
+impl From<InputEvent> for SysInputEvent {
+    fn from(value: InputEvent) -> Self {
+        Self {
+            __bindgen_anon_1: value.sequence.into(),
+            key: value.key.into(),
+            type_: value.r#type.into(),
+        }
+    }
+}
+
+/// An error which may occur while trying
+/// to convert raw [`SysInputEvent`] to [`InputEvent`].
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub enum FromSysInputEventError {
+    InvalidKey(FromSysInputKeyError),
+    InvalidType(FromSysInputTypeError),
+}
+
+impl From<FromSysInputKeyError> for FromSysInputEventError {
+    fn from(value: FromSysInputKeyError) -> Self {
+        Self::InvalidKey(value)
+    }
+}
+
+impl From<FromSysInputTypeError> for FromSysInputEventError {
+    fn from(value: FromSysInputTypeError) -> Self {
+        Self::InvalidType(value)
+    }
+}
+
+/// [`InputEvent`] sequence.
+///
+/// This is a transparent view over [`u32`](prim@u32) with the following representation:
+///
+/// | Bits    | 31..30 | 29..0  |
+/// |---------|--------|--------|
+/// | Payload | Source | Counter|
+///
+/// Corresponds to raw [`SysInputEventSequence`].
+///
+/// # Example usage
+///
+/// Decoding a raw `u32` value:
+///
+/// ```
+/// use flipperzero::input::InputEventSequence;
+/// let sequence = InputEventSequence::from(0b10__000000_10101010_11110000_11111111u32);
+/// assert_eq!(0b10, sequence.source());
+/// assert_eq!(0b10101010_11110000_11111111, sequence.counter());
+/// ```
+#[repr(transparent)]
+#[derive(Copy, Clone, Debug, uDebug, Hash, Eq, PartialEq, PartialOrd, Ord)]
+pub struct InputEventSequence(u32);
+
+impl InputEventSequence {
+    pub const fn source(self) -> u8 {
+        ((self.0 >> 30) & 0b11) as u8
```

Comment:
It might be nice to define these as `SOURCE_MASK` and `SOURCE_SHIFT` constants.

Reply:
All done